🐞 Implemented debug view

DarkMessiah opened the pull request:
## Addressed issues

Implemented: https://github.com/bitbrain/beehave/issues/1

## Screenshots


https://user-images.githubusercontent.com/1756388/218790716-c9bfb2c5-ea4a-49c9-a1d2-3d562a05c7bf.mp4

DarkMessiah committed:
🐞 Implemented debug view

bitbrain commented on addons/beehave/debug/debugger.gd:
```diff
@@ -0,0 +1,58 @@
+
+func _capture(message: String, data: Array, session_id: int) -> bool:
+	if message == "beehave:register_tree":
+		debbuger_tab.register_tree(data[0])
```
minor: rename `debbuger_tab` -> `debugger_tab`
```suggestion
		debugger_tab.register_tree(data[0])
```

bitbrain commented on addons/beehave/debug/debugger.gd:
```diff
@@ -0,0 +1,58 @@
+	session.started.connect(debbuger_tab.start)
+	session.stopped.connect(debbuger_tab.stop)
+
+	debbuger_tab.name = "🐝Beehave"
```
@DarkMessiah does Godot already support emojis in the editor? I remember previously that it had difficulties rendering it.

DarkMessiah replied:
Don't see any problem right now. I can remove emoji :)

bitbrain replied:
Leave it in for now, I like emojis!

bitbrain commented:
Well done! I will first aim to merge #126 (which may result in some conflicts here)

bitbrain commented on addons/beehave/debug/graph_edit.gd:
```diff
@@ -0,0 +1,219 @@
+	updating_graph = false
+
+
+func _add_nodes(node: Dictionary) -> void:
```
It is not guaranteed that `node` has any data (especially when switching scenes). When I ran it while switching scenes it gave me an error that `node` does not contain things like `name`. I solved the issue by doing this at the start of this method:
```suggestion
func _add_nodes(node: Dictionary) -> void:
   if node.is_empty():
	return
```

bitbrain commented on addons/beehave/nodes/beehave_tree.gd:
```diff
@@ -149,7 +155,22 @@ func _exit_tree() -> void:
+
+
+func _get_debugger_data(node: Node) -> Dictionary:
+	var data := { path = node.get_path(), name = node.name, type = node.get_class_name(), id = str(node.get_instance_id()) }
```
It seems that this method is not executed only on this node (`BeehaveTree`) but on **every** node. For example, when running it on my game it complains that some completely unrelated node does not have the method called `get_class_name()`. This might be a bug with Godot? Anyways, I solved this by adding the following infront of the method:
```suggestion
	if not node is BeehaveTree and not node is BeehaveNode:
		return {}
	var data := { path = node.get_path(), name = node.name, type = node.get_class_name(), id = str(node.get_instance_id()) }
```

bitbrain reviewed (changes requested):
I tested this change now in my game and it works quite nicely, apart from some smaller hickups. For larger trees, it becomes a bit difficult to follow what is going on, perhaps there is a way to better position the nodes to the space is better utilised?

![Capture](https://user-images.githubusercontent.com/822035/219130869-4d0831b6-fd2d-490e-80be-a0e7a64261b6.PNG)

I can see that in larger trees it can become a horizontal scroll fest and perhaps we can "spread out" the tree a bit more space-wise? If this is too difficult to do now do not worry - I am happy for us to write up a follow-up ticket to improve the positioning.

![Capture](https://user-images.githubusercontent.com/822035/219133686-872534c4-5ad8-4a76-b475-c58329763cee.PNG)

DarkMessiah commented:
* Added floating window feature.
* Removed ICONS dictionary. Icons are taken from `global_class_list`.
* Removed stylebox resources. Now all styleboxes are generated from GraphNode theme with the right scale.
* Fixed nodes positioning.
* Moved ports icons to `icons` directory.

bitbrain commented on addons/beehave/debug/debugger.gd:
```diff
@@ -0,0 +1,116 @@
+extends EditorDebuggerPlugin
+
+const DebuggerTab := preload("debugger_tab.gd")
+var debugger_tab := DebuggerTab.new()
```
Where do we free this node? I remember that anything allocated with `.new()` needs be called `.free()` manually.

bitbrain commented on addons/beehave/debug/graph_node.gd:
```diff
@@ -0,0 +1,133 @@
+@tool
+extends GraphNode
+
+const DEFAULT_COLOR := Color("#dad4cb")
```
Not for this PR but something to think about in future: do we want to specify the "beehave colours" in some sort of theme resource and ship it with the addon? These colours could then be accessed from that theme directly.

bitbrain committed:
Merge branch 'godot-4.x' into debug_view

bitbrain commented:
@DarkMessiah thank you for your contribution.
